pkg/workload/kv/kv.go

```diff
@@ -942,40 +965,90 @@ func (g *sequentialGenerator) state() string {
 	return fmt.Sprintf("S%d", g.seq.read())
 }
 
-type zipfGenerator struct {
-	seq    *sequence
+type writeSequenceSource struct {
+	*sequence
+}
+
+func (wss writeSequenceSource) Int63() int64 {
+	return int64(wss.Uint64() & 0x7fffffffffffffff) // Masks high bit
+}
+
+func (wss writeSequenceSource) Uint64() uint64 {
+	hasher := crc64.New(crc64.MakeTable(crc64.ECMA))
+	buf := []byte{0, 0, 0, 0, 0, 0, 0, 0}
+	// only increment the sequence during write calls
+	binary.BigEndian.PutUint64(buf[:8], uint64(wss.write()))
+	_, _ = hasher.Write(buf[:8])
+	return hasher.Sum64()
+}
+
+func (wss writeSequenceSource) Seed(seed int64) {
+	wss.val.Store(seed)
+}
+
+type readSequenceSource struct {
+	*sequence
 	random *rand.Rand
-	zipf   *zipf
 }
 
-// Creates a new zipfian generator.
-func newZipfianGenerator(seq *sequence, rng *rand.Rand) *zipfGenerator {
-	return &zipfGenerator{
-		seq:    seq,
-		random: rng,
-		zipf:   newZipf(1.1, 1, uint64(math.MaxInt64)),
+func (rss readSequenceSource) Int63() int64 {
+	return int64(rss.Uint64() & 0x7fffffffffffffff) // Masks high bit
+}
+
+func (rss readSequenceSource) Uint64() uint64 {
+	hasher := crc64.New(crc64.MakeTable(crc64.ECMA))
+	buf := []byte{0, 0, 0, 0, 0, 0, 0, 0}
+	// Unlike in the writeSequenceSource, fetch the last value in the sequence
+	// (don't increment the sequence).
+	v := rss.read()
+	if v != 0 {
+		// To prevent an immediate read of a recently written value, find a random
+		// value between [0,v) to read.
+		v = rss.random.Int63n(v)
 	}
+	binary.BigEndian.PutUint64(buf[:8], uint64(v))
+	_, _ = hasher.Write(buf[:8])
+	return hasher.Sum64()
 }
 
-// Get a random number seeded by v that follows the
-// zipfian distribution.
-func (g *zipfGenerator) zipfian(seed int64) int64 {
-	randomWithSeed := rand.New(rand.NewSource(seed))
-	return int64(g.zipf.Uint64(randomWithSeed))
+func (rss readSequenceSource) Seed(seed int64) {
+	// noop
+}
+
+type zipfGenerator struct {
+	writeZipf *rand.Zipf
+	readZipf  *rand.Zipf
+	random    *rand.Rand
+	seq       *sequence
+}
+
+func newZipfianGenerator(
+	seq *sequence, rng *rand.Rand, zipfianS float64, zipfianV float64,
+) *zipfGenerator {
+	// {read,write}SequenceSources share the same sequence.  readSequenceSource
+	// never mutates the sequence, whereas writeSequenceSource does increments the
+	// sequence.
+	readSS := readSequenceSource{
+		sequence: seq,
+		random:   rng,
+	}
+	writeSS := writeSequenceSource{
+		sequence: seq,
+	}
+	return &zipfGenerator{
+		writeZipf: rand.NewZipf(rand.New(writeSS), zipfianS, zipfianV, uint64(math.MaxInt64)),
```

Comment:
`rand/v2/NewZipf` takes advantage of more efficient `Float64`; why not v2?

Reply:
It's not directly API compatible with `math/rand` ... hold on, let me do that real quick.